Move to Maven/Tycho for projects and build infrastructure

LorenzoBettini opened the pull request:
Replaces https://github.com/eclipse/xtext-lib/pull/500

LorenzoBettini committed:
xtext-dev-bom as a Maven project

LorenzoBettini committed:
parent POM with starting target platform

LorenzoBettini committed:
ignore target

LorenzoBettini committed:
.xbase.lib compiles

LorenzoBettini committed:
.xbase.lib Maven/Tycho project

LorenzoBettini committed:
build .xbase.lib

LorenzoBettini committed:
fixed version in BOM

LorenzoBettini committed:
GitHub Actions workflow

LorenzoBettini committed:
xtend.lib.macro compiles

LorenzoBettini committed:
.xtend.lib.macro as Maven/Tycho project

LorenzoBettini committed:
build .xtend.lib.macro

LorenzoBettini committed:
.xtend.lib compiles

LorenzoBettini committed:
.xtend.lib as Maven/Tycho project

LorenzoBettini committed:
build .xtend.lib

LorenzoBettini committed:
guava-testlib in TP as Maven artifact

LorenzoBettini committed:
.xbase.lib.tests compiles

LorenzoBettini committed:
.xtend.lib.tests as Maven/Tycho project

LorenzoBettini committed:
build xbase.lib.tests, tests are green

LorenzoBettini committed:
xtend.lib.tests compiles

LorenzoBettini committed:
xtend.lib.tests Maven/Tycho project

LorenzoBettini committed:
build xtend.lib.tests, tests are green

LorenzoBettini committed:
xbase.lib.gwt removed gradle nature

LorenzoBettini committed:
added POM to xbase.lib.gwt

taking inspiration from
https://github.com/eclipse/xtext/tree/v2.10.0/plugins/org.eclipse.xtext.xbase.lib.gwt

LorenzoBettini committed:
xbase.lib.gwt as Maven/tycho project

LorenzoBettini committed:
xbase.lib.gwt removed gradle nature

LorenzoBettini committed:
xtend.lib.gwt as Maven/tycho project

LorenzoBettini committed:
.xtend.lib.gwt compiles as Maven project

LorenzoBettini committed:
fixed xtend.lib.gwt.test

LorenzoBettini committed:
build xbase.lib.feature

LorenzoBettini committed:
build xtend.lib.feature

LorenzoBettini committed:
added names to features' POMs

LorenzoBettini committed:
p2 repository

LorenzoBettini committed:
removed old releng directory

LorenzoBettini committed:
updated Jenkins build

LorenzoBettini committed:
removed features from the p2 repository

LorenzoBettini commented:
@cdietrich After removing the features the xtext-eclipse still fails.
I just realized that .xbase.lib (without .feature) is what we actually publish: https://download.eclipse.org/modeling/tmf/xtext/updates/releases/2.29.0/features/

That's a small disgrace because changing names now would be disruptive.

I'll see whether with the suggestion https://github.com/eclipse/xtext-lib/pull/500#issuecomment-1387173352 but I don't like it right from the start...

If also that solution does not work, I can still fix things: it's just a matter of building the features (because we do the same "mistake" in most of our features) and the p2 repository on a separate Maven reactor (i.e., on another Maven build step).

Changing the eclipse project's name, on the other hand, can be done without problems.

cdietrich commented:
i dont know if we want to change the feature name. does it work to add the feature.group to pom and project but keep the id?
i hoped @szarnekow has an opinion on that

cdietrich commented:
also the category.xml does not contain it anymore

cdietrich commented:
of course we could also rename the two features and adopt all downstream stuff.
and add it to the release notes

@szarnekow what do you think?

cdietrich commented:
https://download.eclipse.org/eclipse/updates/4.27-I-builds/I20230118-0200/plugins/
and
https://download.eclipse.org/eclipse/updates/4.27-I-builds/I20230118-0200/features/
has both a `org.eclipse.jdt_`
wonder how they do that.
they have a different group-id
https://github.com/eclipse-jdt/eclipse.jdt/blob/701de3cca1ea57ea701a937fd27426757c6a5dd2/org.eclipse.jdt-feature/pom.xml#L19

LorenzoBettini commented:
> also the category.xml does not contain it anymore

it's because in my latest commit I removed it

LorenzoBettini commented:
> https://download.eclipse.org/eclipse/updates/4.27-I-builds/I20230118-0200/plugins/ and https://download.eclipse.org/eclipse/updates/4.27-I-builds/I20230118-0200/features/ has both a `org.eclipse.jdt_` wonder how they do that. they have a different group-id https://github.com/eclipse-jdt/eclipse.jdt/blob/701de3cca1ea57ea701a937fd27426757c6a5dd2/org.eclipse.jdt-feature/pom.xml#L19

uh! I haven't thought about a different groupId!!!
I'll try that ASAP!

LorenzoBettini committed:
Revert "removed features from the p2 repository"

This reverts commit 8f2a288c895f2f8f8d13c897b2906aa0c8b57d92.

szarnekow commented:
> of course we could also rename the two features and adopt all downstream stuff.
and add it to the release notes

If it keeps things simple, I'd be in favor of fixing the names, make them more consistent, and announce the change.

Most clients will consume the artifacts as plugins anyhow.

LorenzoBettini committed:
updated project configuration

LorenzoBettini committed:
changed groupId of features

Because they have the same artifactId of bundles

LorenzoBettini commented:
@cdietrich it works! How can it be I haven't thought about that :)
When the Jenkins build has finished we can retrigger the job for xtext-eclipse

LorenzoBettini commented:
@cdietrich I retriggered the xtext-eclipse job and the TP is being resolved...

LorenzoBettini commented:
> @cdietrich I retriggered the xtext-eclipse job and the TP is being resolved...

https://ci.eclipse.org/xtext/job/xtext-eclipse/job/lb_2052_maven_tycho/ builds again!

LorenzoBettini commented:
> > of course we could also rename the two features and adopt all downstream stuff.
> > and add it to the release notes
> 
> If it keeps things simple, I'd be in favor of fixing the names, making them more consistent, and announcing the change.
> 
> Most clients will consume the artifacts as plugins anyhow.

So it looks like many Eclipse projects do have features with the same id of bundles (JDT and, IIRC, EMF), so we could also live with that. From the Maven point of view, it's just a matter of having a different groupId, which is a clean solution.

What I don't like of the current shape of Xtext features, like .xbase.lib feature, is that they include bundles XXX and XXX.source. This has also always sounded odd. For sources there are source features. I have already performed such a change in this PR. So now we have features and source features:

![image](https://user-images.githubusercontent.com/1202254/213253321-45a88c8d-cbd4-4a28-b333-47a481c8dbb9.png)

Are you OK with that?

The thing that might sounds strange (though again you find in several Eclipse projects) is that by default Tycho adds the string "Developer Resources" to the generated feature, e.g., 

```Xtend Library Developer Resources```

We can customize that. Do you prefer simply "Sources"? E.g.,

```Xtend Library Sources```

cdietrich commented:
i somewhere read in my platform, tycho, cross-projects whatever mails that source features should be abandoned

LorenzoBettini committed:
xbase.lib.tests is a pure Maven project

No need to have guava-testlib in the target platform, since we consume
it as a Maven dependency

LorenzoBettini commented:
> i somewhere read in my platform, tycho, cross-projects whatever mails that source features should be abandoned

@cdietrich you're referring to this maybe? https://www.eclipse.org/lists/cross-project-issues-dev/msg19553.html

If I understand it correctly Mikael means that we won't have to care about them because everything should be automatic. Another good motivation NOT to include source bundles in the main features.

LorenzoBettini commented:
Concerning this previous note:

> * xbase.lib.tests is not even an Eclipse plug-in project
>   transformed in a plug-in project and added this to the TP:
> 
> ```
> 		<location includeDependencyDepth="none" includeDependencyScopes="compile" includeSource="true" missingManifest="generate" type="Maven">
> 			<dependencies>
> 				<dependency>
> 					<groupId>com.google.guava</groupId>
> 					<artifactId>guava-testlib</artifactId>
> 					<version>30.1-jre</version>
> 					<type>jar</type>
> 				</dependency>
> 			</dependencies>
> 			<instructions><![CDATA[
> Bundle-Name:           Bundle derived from maven artifact ${mvnGroupId}:${mvnArtifactId}:${mvnVersion}
> version:               ${version_cleanup;${mvnVersion}}
> Bundle-SymbolicName:   ${mvnGroupId}.${mvnArtifactId}
> Bundle-Version:        ${version}
> Import-Package:        *;resolution:=optional
> Export-Package:        *;version="${version}";-noimport:=true
> DynamicImport-Package: *
> ]]></instructions>
> 		</location>
> ```
> 
> because `guava-testlib` is not in Orbit. I don't know if Oomph will be able to handle that. The alternative is to turn it into a plain Maven project and take the dependency directly from Maven central. I can investigate.

That is not needed anymore: .xbase.lib.tests is now a pure Maven project so we take everything from Maven central! :)

cdietrich commented:
@LorenzoBettini yes this is the message i meant
i understand it: dont provide source features, just include the sources in the p2 repo and tycho will help with that

LorenzoBettini committed:
put comment in the right place in the POM

LorenzoBettini committed:
eclipse-jarsigner-plugin

LorenzoBettini committed:
eclipse-sign profile in Jenkins

LorenzoBettini committed:
separate projects for p2 and maven

LorenzoBettini committed:
harmonized parent with xtext-core parent

LorenzoBettini committed:
harmonized parent with xtext-extras parent

LorenzoBettini committed:
harmonized parent with xtext-xtend parent

LorenzoBettini committed:
prepared sonatype-oss-release profile

LorenzoBettini committed:
distributionManagement

LorenzoBettini committed:
prepared Jenkinsfile for Maven deploy

LorenzoBettini committed:
skipTests in profile sonatype-oss-release

LorenzoBettini committed:
reference to Maven releng project

LorenzoBettini committed:
distributionManagement in dev-bom

because it does not inherit from the parent

cdietrich commented:
@LorenzoBettini  please note:

we use https://oss.sonatype.org/content/repositories/snapshots/ for snapshots
and https://s01.oss.sonatype.org/service/local/staging/deploy/maven2/ for milestones and releases

see also https://github.com/xtext/publishing/blob/aa97f2b592e9b64debde904ce262c35514a296a2/buildSrc/src/main/java/io/typefox/publishing/MavenUploadRepository.java#L13

LorenzoBettini commented:
> @LorenzoBettini please note:
> 
> we use https://oss.sonatype.org/content/repositories/snapshots/ for snapshots and https://s01.oss.sonatype.org/service/local/staging/deploy/maven2/ for milestones and releases
> 
> see also https://github.com/xtext/publishing/blob/aa97f2b592e9b64debde904ce262c35514a296a2/buildSrc/src/main/java/io/typefox/publishing/MavenUploadRepository.java#L13

Thanks! I forgot to ask. I'll update the URL for release.

However, the snapshot repository is the same I've used, and sonatype is happy :)

https://oss.sonatype.org/content/repositories/snapshots/org/eclipse/xtext/xtext-dev-bom/2.30.0-SNAPSHOT/

For the moment, I created a TEMP build job on JIRO, using a dedicated Jenkinsfile `jenkins/Jenkinsfile-maven-deploy`

LorenzoBettini committed:
updated distributionManagement release repository

cdietrich commented:
we use 2 different repos currently

LorenzoBettini commented:
> we use 2 different repos currently. i dont know if and how they also sync snapshots among them

This is the standard section for deployment of Maven artifacts:

```xml
	<distributionManagement>
		<snapshotRepository>
			<id>ossrh</id>
			<url>https://oss.sonatype.org/content/repositories/snapshots/</url>
		</snapshotRepository>
		<repository>
			<id>ossrh</id>
			<url>https://s01.oss.sonatype.org/service/local/staging/deploy/maven2/</url>
		</repository>
	</distributionManagement>
```

did you mean something else? Maven knows by itself if it's a snapshot (-SNAPSHOT in the version) or something that goes into the release repository (no -SNAPHSOT in the version).

LorenzoBettini committed:
removed wrong configuration option of javadoc plugin

cdietrich commented:
yes this is what i meant

LorenzoBettini committed:
eclipse-sign also when deploying Maven artifacts

LorenzoBettini commented:
Looks like signing when also running tests is likely to break tests:

```
14:24:40  [ERROR] class "org.eclipse.xtext.parsetree.reconstr.SerializationBug269362Test"'s signer information does not match signer information of other classes in the same package
14:24:40  [ERROR] org.apache.maven.surefire.booter.SurefireBooterForkException: There was an error in the forked process
14:24:40  [ERROR] class "org.eclipse.xtext.parsetree.reconstr.SerializationBug269362Test"'s signer 
```

Since I'd avoid creating other Jenkins jobs, I'll execute tests in the first run, and then execute another run skipping tests doing the actual signing. After all, these are just experiments.

LorenzoBettini commented:
mh... no, that doesn't work:

- if I run the build with signing and then the build with tests, I lose the archived p2 repo with signed contents
- if I do it the other way around, I lose the test results to archive

For these experiments, I have to create separate jobs just for archiving the p2 repository with signed contents. I'll use these latter p2 repositories to create later an aggregated p2 repository with signed contents in the umbrella.

This will take me some more time...

cdietrich commented:
is the jar first built and then tested? or is target/classes used in tests?

LorenzoBettini commented:
That depends on whether the tests are run with maven-surefire (before package) or with tycho-surefire (after package). In any case, it's not worthwhile to sign in the same build where we test. I thought I could do the two things together but that doesn't work.

LorenzoBettini commented:
> so we have to run all builds twice. or can we maven sign separately? by calling deploy jar in a loop or so

I'm creating build jobs unrelated from the multi-branch pipeline, with a dedicated Jenkinsfile used only for Maven deploy and p2 deploy. One for Maven deploy and one for p2 deploy. A pair for each Git repository.

This is just to test that sonatype is happy with our deployed Maven artifacts and to create p2 repositories with signed contents that I can aggregate into the main p2 repository.

These are all temporary workarounds due to the split Git repositories.

However, in the end, it's still better to have an on-demand job for Maven deploy and another one for p2 release.

LorenzoBettini commented:
`xtext-core` Maven artifacts are accepted by sonatype, e.g., https://oss.sonatype.org/content/repositories/snapshots/org/eclipse/xtext/org.eclipse.xtext/2.30.0-SNAPSHOT/

temporary p2 repository for xtext-core with signed contents is here https://ci.eclipse.org/xtext/view/Xtext-release/job/TEMP-xtext-core-p2-deploy/lastSuccessfulBuild/artifact/build/p2-repository/

cdietrich commented:
are the criteria for sonatype to accept snapshots and the releases the same.
i vaguely remember there was some differences

LorenzoBettini commented:
> mh... no, that doesn't work:
> 
>     * if I run the build with signing and then the build with tests, I lose the archived p2 repo with signed contents
> 
>     * if I do it the other way around, I lose the test results to archive
> 
> 
> For these experiments, I have to create separate jobs just for archiving the p2 repository with signed contents. I'll use these latter p2 repositories to create later an aggregated p2 repository with signed contents in the umbrella.
> 
> This will take me some more time...

I forgot to mention that `xtext-lib` (the first one I started from with these experiments) does not have such problems: we can sign and run tests in the same build, because the tests are in the same project containing main source folders.

szarnekow commented:
So the problem is caused by split packages between production and test projects?

LorenzoBettini commented:
> So the problem is caused by split packages between production and test projects?

That's my understanding as well.
However, as I said, from what I know, you sign only when releasing (snapshots, milestones, etc.), where you typically don't run tests.

LorenzoBettini commented:
`xtext-extras` Maven artifacts are accepted by sonatype

temporary p2 repository https://ci.eclipse.org/xtext/view/Xtext-release/job/TEMP-xtext-extras-p2-deploy/lastSuccessfulBuild/artifact/build/p2-repository/

cdietrich commented:
> However, as I said, from what I know, you sign only when releasing (snapshots, milestones, etc.), where you typically don't run tests.
we currently do

there is another "something that rings my mind"
- maven surefire vs tycho tests using maven surefire. but i cannot relate right now
- maybe phases test vs integ test

LorenzoBettini commented:
> > However, as I said, from what I know, you sign only when releasing (snapshots, milestones, etc.), where you typically don't run tests.
> > we currently do
> 
> there is another "something that rings my mind"
> 
>     * maven surefire vs tycho tests using maven surefire. but i cannot relate right now
> 
>     * maybe phases test vs integ test

Yes, I have already answered above :) https://github.com/eclipse/xtext-lib/pull/501#issuecomment-1406557481

what do you mean by "we currently do"?

cdietrich commented:
https://github.com/eclipse/xtext/blob/master/Builds.md#preparing-milestones-and-releases
we run tests during the release process.
we run a normal build

LorenzoBettini commented:
> https://github.com/eclipse/xtext/blob/master/Builds.md#preparing-milestones-and-releases we run tests during the release process. we run a normal build

It's a waste of time running tests during release ;)

cdietrich commented:
we have some utils that behave differently based on how version numbers look like ....
and thus test fails/bugs in the past

LorenzoBettini commented:
> we have some utils that behave differently based on how version numbers look like .... and thus test fails/bugs in the past
> 
> [eclipse/xtext-core#1351](https://github.com/eclipse/xtext-core/issues/1351)

nothing prevents you from running the tests with what you want to release **before** you release, and then release skipping tests.

cdietrich commented:
so we are back at running the build twice?

szarnekow commented:
Not that a `mvn deploy` without `clean` does not necessarily run the build twice. If things go well, the produced JARs would just be picked up, signed and deployed, won't they?

LorenzoBettini commented:
> Not that a `mvn deploy` without `clean` does not necessarily run the build twice. If things go well, the produced JARs would just be picked up, signed and deployed, won't they?

yes, though we should skip also xtend-maven-plugin in that case.

In any case, I don't see a huge problem here: if you want to make sure the current version works you run the tests, when the tests are green, you run the build skipping the tests: it's not a big deal ;)

cdietrich commented:
> In any case, I don't see a huge problem here: if you want to make sure the current version works you run the tests, when the tests are green, you run the build skipping the tests: it's not a big deal ;)

my goal would be to have a one click solution (as we have it right now)
and not starting 4+ jenkins jobs manually

LorenzoBettini commented:
I haven't said that: when I talk about a build I mean a maven command. Nothing prevents us to run several maven commands one after the other in the same job.

The experiments I'm doing now are different because I first need to make sure each single repo does the right thing.

In general, though, it's better to have a clean, maintainable and simple configuration that might require two clicks (two jobs) than a complex over-enginereed one requiring one click ;)

For sure, it's better to have different jobs for everyday development builds and for releasing.

LorenzoBettini commented:
> > so we have to run all builds twice. or can we maven sign separately? by calling deploy jar in a loop or so
> 
> I'm creating build jobs unrelated from the multi-branch pipeline, with a dedicated Jenkinsfile used only for Maven deploy and p2 deploy. One for Maven deploy and one for p2 deploy. A pair for each Git repository.
> 
> This is just to test that sonatype is happy with our deployed Maven artifacts and to create p2 repositories with signed contents that I can aggregate into the main p2 repository.
> 
> These are all temporary workarounds due to the split Git repositories.
> 
> However, in the end, it's still better to have an on-demand job for Maven deploy and another one for p2 release.

I think I found a solution that allows me to run the tests and later to sign in the same Jenkins job. The trick is

- don't build the p2 repository during the first run when we run the tests
- on the second run, don't run the tests, don't `clean`, but sing and create the p2 repo

This way, in the same Jenkins job I can record test results and archive the p2 repository with signed contents.

I'm not completely happy with this solution, but at least, it's the cleanest one that saves time, and it's temporary anyway. I'll apply that also to the other git repositories.

However, in the meantime, @merks , when you have time, could you please check that also this p2 repo is correct? https://ci.eclipse.org/xtext/job/xtext-core/job/lb_2052_maven_tycho/lastSuccessfulBuild/artifact/build/p2-repository/

merks commented:
It still looks fine:

https://download.eclipse.org/oomph/archive/reports-extra/xtext-tycho/ci.eclipse.org/xtext/job/xtext-core/job/lb_2052_maven_tycho/lastSuccessfulBuild/artifact/build/p2-repository/index.html

szarnekow commented on .gitignore:
```diff
@@ -6,3 +6,5 @@ build/
 *.xtendbin
 org.eclipse.buildship.core.prefs
 .DS_Store
+
```
Nit: blank line surprises me.

szarnekow commented on 2-tycho-build.sh:
```diff
@@ -4,11 +4,12 @@ if [ -z "$JENKINS_URL" ]; then
 mvn \
-  -f releng \
+  -f org.eclipse.xtext.p2.releng \
+  clean verify \
```
I'd expect the goals to be at the end of the argument list. Why did you move them up?

LorenzoBettini replied:
I usually prefer to have the goals initially, but I can move them down. Do you prefer that?

szarnekow replied:
Both variants are fine for me. It was more curiosity than a real thing.

szarnekow commented on 3-maven-build.sh:
```diff
@@ -0,0 +1,17 @@
+mvn \
+  -f org.eclipse.xtext.maven.releng \
+  -DskipTests=true \
+  clean deploy \
```
Do you really want to clean in step 3?

LorenzoBettini replied:
I don't think so. I'll remove that.

szarnekow commented on org.eclipse.xtend.lib.gwt.test/.settings/org.eclipse.core.resources.prefs:
```diff
@@ -1,2 +1,3 @@
 eclipse.preferences.version=1
 encoding/<project>=ISO-8859-1
+encoding/src=ISO-8859-1
```
Curiosity: Isn't this redundant?

LorenzoBettini replied:
Looks like m2e likes to add that information for source folders when you update the project.

szarnekow commented on org.eclipse.xtext.xbase.lib.gwt/pom.xml:
```diff
@@ -0,0 +1,145 @@
+
```
Nit: blank line at the top of the file

LorenzoBettini replied:
OK

szarnekow commented on org.eclipse.xtext.xbase.lib.gwt/pom.xml:
```diff
@@ -0,0 +1,145 @@
+	<dependencies>
+		<dependency>
+			<groupId>com.google.code.findbugs</groupId>
+			<artifactId>jsr305</artifactId>
```
Is this a transitive dependency? Can we make it compile-only?

LorenzoBettini replied:
In that project, I guess it's a copy and paste mistake, because the old build.gradle file did not mention that.
I'll try to remove that and see if everything still works.

LorenzoBettini replied:
It was present also in xtend.lib.gwt. Looks like it's not needed at all (locally the build is green). I removed it from both projects: https://github.com/eclipse/xtext-lib/pull/501/commits/43171e0bcfdab7c5f35aaa06f501ab3741695026

szarnekow commented on org.eclipse.xtext.xbase.lib.tests/.settings/org.eclipse.core.resources.prefs:
```diff
@@ -1,3 +1,4 @@
-#Tue Aug 18 18:37:57 CEST 2009
 eclipse.preferences.version=1
 encoding/<project>=ISO-8859-1
+encoding/src=ISO-8859-1
```
Same question as above: redundant?

szarnekow commented on pom.xml:
```diff
@@ -0,0 +1,570 @@
+		<developer>
+			<name>Sebastian Zarnekow</name>
+			<email>sebastian.zarnekow@gmail.com</email>
+			<organization>itemis</organization>
```
I'm not affiliated with itemis. Please remove.

LorenzoBettini replied:
should I put "independent"?
That comes from the current contents of the generated POMs, so I'll have to update that also in the other projects.

szarnekow replied:
Please remove the organization. If it's mandatory, please put `Sebastian Zarnekow Consulting`

szarnekow commented on org.eclipse.xtext.dev-bom/pom.xml:
```diff
@@ -0,0 +1,482 @@
+		<developer>
+			<name>Sebastian Zarnekow</name>
+			<email>sebastian.zarnekow@gmail.com</email>
+			<organization>itemis</organization>
```
I'm not affiliated with itemis at this time. Please remove the organization information.

LorenzoBettini committed:
renamed maven aggregator artifactId

LorenzoBettini committed:
removed blank line

LorenzoBettini committed:
removed organization for Sebastian

LorenzoBettini committed:
removed dep com.google.code.findbugs:jsr305

cdietrich commented on org.eclipse.xtext.xbase.lib.gwt/pom.xml:
```diff
@@ -14,10 +14,6 @@
 	<description>The runtime library for the Xbase programming language.</description>
 
 	<dependencies>
-		<dependency>
```
isnt this transitive dependency?

LorenzoBettini replied:
Right!

![image](https://user-images.githubusercontent.com/1202254/215452428-92fa3881-81d5-4ecf-b435-32b33d97d4c0.png)

And, slightly OT, IIRC, one that gives a lot of headaches in testing due to mismatch in signatures with javax.annotations. When testing with mockito for example. ;)

LorenzoBettini commented:
Looks like JIRO is overloaded (due to the new milestone release?).

@cdietrich, concerning the new milestone, I'll try to update my branches, but first, I'd like to conclude the "harmonization" of all the Git repositories to produce a complete p2 repository with signed contents.

cdietrich commented:
@LorenzoBettini @szarnekow we need to discuss how we bring that  live with 2.31.0.M1 anyway
i wonder if we should PR the manifest/build.properties + rename/refactor changes to master independently
and then with the real stuff PR the pom/nature etc changes

LorenzoBettini commented:
Some updates:

- Xtend p2 repository with signed contents is now available from here https://ci.eclipse.org/xtext/job/xtext-xtend/job/lb_2052_maven_tycho/lastSuccessfulBuild/artifact/build/p2-repository/
- Xtend Maven artifacts (snapshots) are accepted by sonatype
- `xtext-eclipse` p2 repository with signed contents is now available from here: https://ci.eclipse.org/xtext/job/xtext-eclipse/job/lb_2052_maven_tycho/lastSuccessfulBuild/artifact/build/p2-repository/

@cdietrich just as a confirmation: excluding `xtend-maven-plugin`, which is taken with the current latest released versions, all the other Maven artifacts dependencies concerning our bundles, even across Git repositories, should be taken with -SNAPSHOT version, right? By default they would be taken from sonatype snapshots. In the build are taken from latest Jenkins Maven repositories, is that correct?

That is, `<xtext-current-release-version>2.30.0.M1</xtext-current-release-version>` should be in all Git repositories `<xtext-current-release-version>${project.version}</xtext-current-release-version>`. Am I correct?

cdietrich commented:
yes. the same as it would be a monorepo. the only place we use **other version** is xtend-maven-plugin.
and it should be done with profile so that it can be decided if from sonatype or from upstream branches

LorenzoBettini commented:
> yes. the same as it would be a monorepo. the only place we use **other version** is xtend-maven-plugin. and it should be done with profile so that it can be decided if from sonatype or from upstream branches

I'm doing some experiments in `xtext-core` concerning the versioning across Git repositories.

So I introduced `xtext-dev-bom-version` that defaults to `project.version` and can be redefined in projects like xtext.logging that has a different version https://github.com/eclipse/xtext-core/commit/e329fc2f88c92fae965ec959e6bf9cd752d507c9

Besides that, I always use project.version for referring to xtext artifacts coming from other git repositories: https://github.com/eclipse/xtext-core/commit/31358b58febf2dcbd9245c28a1090090baaeec8c

During building, we use the profile `useJenkinsSnapshots` so the snapshots are found. But what about without that profile? For example, in Eclipse? Without snapshot repository specification, snapshots are not found. A profile for sonatype snapshots will not work because `activeByDefault`, as I said in the past, is not really active by default.

Thus, the sonatype snapshot repository is the default (specified outside profiles), and the profile  `useJenkinsSnapshots` explicitly disables that sonatype snapshots repository: https://github.com/eclipse/xtext-core/commit/53f0781e16304cc81e9aeacfe24a1fd970253120

LorenzoBettini committed:
versions-maven-plugin

LorenzoBettini committed:
a profile mentioning releng children

LorenzoBettini committed:
Merge remote-tracking branch 'upstream/master' into lb_2052_maven_tycho

LorenzoBettini committed:
removed old xtext-current-release-version

LorenzoBettini committed:
xtend-maven-plugin-version 2.30.0.M2

LorenzoBettini committed:
harmonized parent POM with other projects

LorenzoBettini committed:
harmonized Jenkins script with other projects

separate maven build, tycho test and tycho sign

LorenzoBettini committed:
coinfigured maven-jar-plugin

LorenzoBettini committed:
configured maven-source-plugin

LorenzoBettini committed:
set execution bit

LorenzoBettini committed:
skip main "compile" goal of xtend-maven-plugin

to avoid generated sources to be considered as main sources by the
maven-compiler-plugin

LorenzoBettini committed:
profile buildP2Repository

LorenzoBettini committed:
Merge remote-tracking branch 'upstream/master' into lb_2052_maven_tycho

LorenzoBettini committed:
configured exec-maven-plugin

LorenzoBettini committed:
added source bundles in features

As it used to be in the original

LorenzoBettini committed:
Merge remote-tracking branch 'origin/master' into lb_2052_maven_tycho

LorenzoBettini committed:
update to lsp4j 0.20.0

LorenzoBettini committed:
disable p2 sign in Jenkinsfile

LorenzoBettini committed:
Merge remote-tracking branch 'upstream/master' into lb_2052_maven_tycho

LorenzoBettini committed:
bumped 2.31.0 in POMs

LorenzoBettini committed:
xtend-maven-plugin-version 2.30.0

LorenzoBettini committed:
updated other eclipse metadata to 2.31.0

LorenzoBettini committed:
removed gradle files

LorenzoBettini committed:
don't sign Maven artifacts in Jenkinsfile

LorenzoBettini committed:
profile for javadoc
